### PR TITLE
Introduce df.spark.repartition

### PR DESCRIPTION
**Pull request by xinrong-databricks**

### CODE REVIEW

File: databricks/koalas/spark/accessors.py [hunk 1]
```diff
@@ -851,6 +851,27 @@ def apply(self, func, index_col: Optional[Union[str, List[str]]] = None) -> "ks.
             )
         return output.to_koalas(index_col)
 
+    def repartition(
+        self, numPartitions: int, index_col: Optional[Union[str, List[str]]] = None
```
Comment (ueshin) [inline databricks/koalas/spark/accessors.py:855]: `num_partitions` instead of `numPartitions`?

File: databricks/koalas/frame.py [hunk 2]
```diff
@@ -9286,6 +9286,13 @@ def gen_names(v, curnames):
         else:
             return DataFrame(internal)
 
+    def repartition(
+        self, numPartitions: Union[int, str], index_col: Optional[Union[str, List[str]]] = None
+    ) -> "ks.DataFrame":
+        return self.spark.repartition(numPartitions, index_col)
+
+    repartition.__doc__ = SparkFrameMethods.repartition.__doc__
+
```
Comment (ueshin) [inline databricks/koalas/frame.py:9295]: I don't think we need to add it to `DataFrame`.

File: databricks/koalas/spark/accessors.py (hunk 1 above)
Comment (ueshin) [reply databricks/koalas/spark/accessors.py:855]: I think this should work without `index_col`, returns the same DataFrame except for partitioning.

File: databricks/koalas/spark/accessors.py [hunk 3]
```diff
@@ -851,6 +851,27 @@ def apply(self, func, index_col: Optional[Union[str, List[str]]] = None) -> "ks.
...
+        numPartitions : int
+            The target number of partitions.
+        index_col: str or list of str, optional, default: None
+            Column names to be used in Spark to represent Koalas' index. The index name
+            in Koalas is ignored. By default, the index is always lost.
+
+        Returns
+        -------
+        DataFrame
+        """
```
Comment (ueshin) [inline databricks/koalas/spark/accessors.py:872]: Could you add simple examples?

File: databricks/koalas/tests/test_dataframe.py [hunk 4]
```diff
@@ -4546,6 +4546,17 @@ def test_explain_hint(self):
         finally:
             sys.stdout = prev
 
+    def test_repartition(self):
+        kdf = ks.DataFrame({"age": [5, 5, 2, 2], "name": ["Bob", "Bob", "Alice", "Alice"]})
+
+        new_kdf = kdf.spark.repartition(7)
+        self.assertEqual(new_kdf.to_spark().rdd.getNumPartitions(), 7)
```
Comment (ueshin) [inline databricks/koalas/tests/test_dataframe.py:4553]: Shall we also test `self.assert_eq(kdf.sort_index(), new_kdf.sort_index())`?

File: databricks/koalas/tests/test_dataframe.py [hunk 5]
```diff
@@ -4546,6 +4546,17 @@ def test_explain_hint(self):
...
+
+        kdf = kdf.set_index("age")
+        new_kdf = kdf.spark.repartition(5)
+        self.assertEqual(new_kdf.to_spark().rdd.getNumPartitions(), 5)
+        self.assertEqual(kdf.index.name, new_kdf.index.name)
```
Comment (ueshin) [inline databricks/koalas/tests/test_dataframe.py:4558]: Could you add tests with mutli-index columns?

Comment (ueshin) [inline databricks/koalas/tests/test_dataframe.py:4558]: Shall we put this test in `test_frame_spark`?

File: databricks/koalas/spark/accessors.py (hunk 3 above)
Comment (xinrong-databricks) [reply databricks/koalas/spark/accessors.py:888]: Certainly! Added.

File: databricks/koalas/tests/test_dataframe.py (hunk 5 above)
Comment (xinrong-databricks) [reply databricks/koalas/tests/test_dataframe.py]: Sounds good!

Comment (xinrong-databricks) [reply databricks/koalas/tests/test_dataframe.py]: Certainly, added.

File: databricks/koalas/tests/test_dataframe.py (hunk 4 above)
Comment (xinrong-databricks) [reply databricks/koalas/tests/test_dataframe.py]: Added!

File: databricks/koalas/spark/accessors.py [hunk 6]
```diff
@@ -851,6 +851,48 @@ def apply(self, func, index_col: Optional[Union[str, List[str]]] = None) -> "ks.
...
+            The target number of partitions.
+
+        Returns
+        -------
+        DataFrame
+
+        Examples
+        --------
+        >>> import databricks.koalas as ks
+        >>> import pyspark
```
Comment (ueshin) [inline databricks/koalas/spark/accessors.py:871]: nit: we don't need to import those.

Comment (xinrong-databricks) [reply databricks/koalas/spark/accessors.py]: Removed.

File: databricks/koalas/spark/accessors.py [hunk 7]
```diff
@@ -851,6 +851,50 @@ def apply(self, func, index_col: Optional[Union[str, List[str]]] = None) -> "ks.
...
+
+        Returns
+        -------
+        DataFrame
+
+        Examples
+        --------
+        >>> kdf = ks.DataFrame({"age": [5, 5, 2, 2],
+        ...         "name": ["Bob", "Bob", "Alice", "Alice"]}).set_index("age")
+        >>> kdf  # doctest: +NORMALIZE_WHITESPACE
```
Comment (ueshin) [inline databricks/koalas/spark/accessors.py:872]: Shall we use `.sort_index()` here as well so that we can see the same as the repartitioned result?

**Review by ueshin — COMMENTED** [review]
Otherwise, LGTM, pending tests.

**Review by HyukjinKwon — APPROVED** [review]
LGTM too. We can also add coalesce I guess :-). Also this should be added into doc.

File: databricks/koalas/spark/accessors.py (hunk 7 above)
Comment (xinrong-databricks) [reply databricks/koalas/spark/accessors.py]: Makes sense! Updated.

**Comment by xinrong-databricks** [comment]
@HyukjinKwon Sounds good! Thank you for your approval!

**Pull request merged**